File: backend/app/api/routes/mastery.py
from uuid import UUID

from fastapi import APIRouter

from app.api.deps import CurrentUser, SessionDep
from app.core.errors import AppErrorException, ErrCode
from app.crud import get_knowledge_point_mastery
from app.domain.mastery import compute_mastery_score, mastery_level
from app.models import KnowledgeMasteryResp, MasteryResp, User

router = APIRouter(prefix="/tasks/children", tags=["mastery"])


@router.get("/{child_id}/mastery", response_model=MasteryResp)
def mastery(
    *, session: SessionDep, user: CurrentUser, child_id: UUID
) -> MasteryResp:
    """查看某娃娃的知识点掌握度看板（F-204 / AC-203）。

    家长可查看自家娃娃；娃娃仅可查看自己的掌握度（双角色分支，
    与 tasks.py 的作答/打卡接口保持一致）。
    """
    if user.role == "child":
        if child_id != user.id:
            raise AppErrorException(ErrCode.TASK_NOT_OWNED, "这不是你的掌握度")
        child = user
    else:  # parent
        child = session.get(User, child_id)
        if child is None or child.parent_id != user.id:
            raise AppErrorException(
                ErrCode.TASK_NOT_YOUR_CHILD, "这不是你家娃娃的掌握度"
            )

    items: list[KnowledgeMasteryResp] = []
    for agg in get_knowledge_point_mastery(session=session, child_id=child_id):
        score = compute_mastery_score(
            total_answers=agg.total_answers,
            correct_answers=agg.correct_answers,
            recent_total=agg.recent_total,
            recent_correct=agg.recent_correct,
            active_wrong=agg.active_wrong,
            max_review_stage=agg.max_review_stage,
        )
        items.append(
            KnowledgeMasteryResp(
                knowledge_point=agg.knowledge_point,
                subject=agg.subject,
                grade=agg.grade,
                total_answers=agg.total_answers,
                correct_answers=agg.correct_answers,
                accuracy=round(agg.correct_answers / agg.total_answers, 2)
                if agg.total_answers
                else 0.0,
                active_wrong=agg.active_wrong,
                max_review_stage=agg.max_review_stage,
                score=score,
                level=mastery_level(
                    total_answers=agg.total_answers,
                    score=score,
                    active_wrong=agg.active_wrong,
                ),
            )
        )
    return MasteryResp(
        child_id=child_id,
        total_knowledge_points=len(items),
        mastered_count=sum(1 for i in items if i.level == "已掌握"),
        items=items,
    )

File: frontend/lib/shared/domain/models/mastery.dart
// 领域模型：与后端 SQLModel schema 对齐的纯 Dart 模型。
// 注意：所有 ID 均为 UUID 字符串（后端用 uuid.UUID）。


class ProgressModel {
  final String childId;
  final int total;
  final int correct;
  final double accuracy;
  final int streakDays;
  final int checkinDays;

  ProgressModel({
    required this.childId,
    required this.total,
    required this.correct,
    required this.accuracy,
    required this.streakDays,
    required this.checkinDays,
  });

  factory ProgressModel.fromJson(Map<String, dynamic> json) {
    return ProgressModel(
      childId: json['child_id'] as String,
      total: json['total'] as int,
      correct: json['correct'] as int,
      accuracy: (json['accuracy'] as num).toDouble(),
      streakDays: json['streak_days'] as int,
      checkinDays: json['checkin_days'] as int,
    );
  }
}

class KnowledgeMasteryModel {
  /// 单个知识点的掌握度（家长看板）。
  final String knowledgePoint;
  final String subject;
  final int grade;
  final int totalAnswers;
  final int correctAnswers;
  final double accuracy;
  final int activeWrong;
  final int maxReviewStage;
  final double score;
  final String level;

  KnowledgeMasteryModel({
    required this.knowledgePoint,
    required this.subject,
    required this.grade,
    required this.totalAnswers,
    required this.correctAnswers,
    required this.accuracy,
    required this.activeWrong,
    required this.maxReviewStage,
    required this.score,
    required this.level,
  });

  factory KnowledgeMasteryModel.fromJson(Map<String, dynamic> json) {
    return KnowledgeMasteryModel(
      knowledgePoint: json['knowledge_point'] as String,
      subject: json['subject'] as String,
      grade: json['grade'] as int,
      totalAnswers: json['total_answers'] as int? ?? 0,
      correctAnswers: json['correct_answers'] as int? ?? 0,
      accuracy: (json['accuracy'] as num?)?.toDouble() ?? 0.0,
      activeWrong: json['active_wrong'] as int? ?? 0,
      maxReviewStage: json['max_review_stage'] as int? ?? 0,
      score: (json['score'] as num?)?.toDouble() ?? 0.0,
      level: json['level'] as String? ?? '',
    );
  }
}

class MasteryModel {
  /// 知识点掌握度看板（家长端）。
  final String childId;
  final int totalKnowledgePoints;
  final int masteredCount;
  final List<KnowledgeMasteryModel> items;

  MasteryModel({
    required this.childId,
    required this.totalKnowledgePoints,
    required this.masteredCount,
    required this.items,
  });

  factory MasteryModel.fromJson(Map<String, dynamic> json) {
    return MasteryModel(
      childId: json['child_id'] as String,
      totalKnowledgePoints: json['total_knowledge_points'] as int? ?? 0,
      masteredCount: json['mastered_count'] as int? ?? 0,
      items: (json['items'] as List? ?? [])
          .map((e) => KnowledgeMasteryModel.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
